GoogleForCreators/web-stories-wp
# Floating Menu: Add Roving Keyboard Navigation

## Context

Floating menu wasn't navigable via keyboard. 

## Summary

Adds the ability to navigate and interact with floating menu inputs via keyboard.

## Relevant Technical Choices

Code's just a _a little_ dirty 🙈  - we're kind of cornered on implementation options without tearing some existing structures apart. I ended up solving this through the following:  

1. Escape hatch for focusable elements in a context menu that should be ignored as `focusableChildren`
2. Checking for the id of the activeElement when the next focusable id comes back as -1 according to `getPrevIndex` (which I extracted from `handleKeyboardNav` in `design-system/../contextMenu/menu` ) so that we can account for focus getting passed back from another popup. 
3. Reusable `FocusTrapButton` in the floating menu elements that wraps around a freeform input (text or numeric), asking for the user to hit 'enter' to focus that element or the group of them and then that watches for 'tab' to resume context menu keyboard navigation. This is definitely me fitting a triangle peg in a circle shaped hole. Added documentation in relevant files since this is weird. 
4. Got rid of the out of date `useGlobalKeyDownEffect` to clear selection and transformations on `esc`. Canvas is in its own focus group now. Getting rid of this enables us to pass focus from the floating menu back to the selected element and continue to traverse the canvas's elements. 

We'd discussed giving the floating menu a role of `toolbar`, doing that would mean we have to also change the children's role from `menuitem` to `group` or `button` and that was just getting messy. The way it exists now, everything is navigable and named. When I run aXe on it I'm only getting issues around the `tab-index: -1` that's added to prevent focus to the menu when the canvas isn't actively selected - it resolves itself when the menu's selected but aXe doesn't know that. If it's felt strongly that the floating menus should be a toolbar then we need to have a new design system component called toolbar and have it be designed as such. I don't feel good forcing that into the context menu. 


## To-do

### Here

- [x] karma tests 
- [x] Code clean up
- [x] Solve whatever is happening with the color picker variants attached to an eyedropper - I can get focus TO them and from them but the `tab` keydowns once I enter the color popup from the context menu popup are getting swallowed. Works fine for border color - not attached to an eyedropper.  
- [x] Add screenshots/demos 
- [x] Fix focus trap button focus on color opacity (only present in shape floating menu - see gif below)

### Separately
- If you use a cursor and select an element, you can get to that element with the keyboard shortcut (ctrl+option+p) but if you esc it to go back to that element, because you didn't start from the canvas, the canvas doesn't know it is focused so it doesn't work. Great if you're only using keyboard and great if you only use the mouse - mixing them, still not 100%. (https://github.com/GoogleForCreators/web-stories-wp/issues/11003)
- I added the floating menu to storybook to make discovery around how to fix this a little faster, it needs more context added to it to really be a mirror copy of what we get in the editor - right now shapes and stickers are the only ones that look how they are supposed to. It might be useful to flesh that out.
- What's the intended keyboard functionality of the eyedropper for color pickers? Feels like a cursor is necessary. Should we block that on keydown enter - just remove it from the keyboard navigation completely? 

## User-facing changes

- color popup on floating menu no longer closes with selection. found that to be really hard to use via keyboard.
- Now you can navigate the floating menus with a keyboard.


| Menu | Demo |
| --- | --- |
| Image | ![focusmenu-image](https://user-images.githubusercontent.com/10720454/163735062-ef8fb09b-2245-4aa4-a56a-2000300e7e6b.gif) |
| Video | ![focusmenu-video](https://user-images.githubusercontent.com/10720454/163735112-649c0378-858a-4cf5-afc1-5dbee550f44e.gif) |
| Text |  ![focusmenu-text](https://user-images.githubusercontent.com/10720454/163735157-77df072f-31c5-478a-bfa7-164ab0926a1f.gif) | 
| Shape 1 | ![focusmenu-shape](https://user-images.githubusercontent.com/10720454/163850793-8fe45896-b01f-4a4c-9518-a5cb5064ba01.gif) | 
| Shape 2 | ![focusmenu-shapewithradius](https://user-images.githubusercontent.com/10720454/163851232-54679995-7b36-48fb-909a-02f6a55e9aab.gif) | 
| Sticker | ![focusmenu-sticker](https://user-images.githubusercontent.com/10720454/163735220-77b5b531-8828-47d8-b95b-1eb27e8820c3.gif) |
| Muli-select | ![focusmenu-multiple](https://user-images.githubusercontent.com/10720454/163735282-75eaf0fa-a309-4550-a544-7269f290d576.gif) | 


## Testing Instructions






This PR can be tested by following these steps:

Add an element to the canvas and use either the cursor or the new keyboard shortcut (`ctrl+option+p`)  and then you're in the floating menu. Everything should be usable via keyboard. Eyedropper is still really weird to use with a keyboard, but that's not new. Check to see if `esc` gets you back to the element, you should be able to use the arrow keys to move the element. One outstanding item is getting the canvas focused again if you didn't start out with just the keyboard. 


## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #11162
